src/ingestion/kafka-queue.ts

```diff
@@ -71,11 +72,20 @@ export class KafkaQueue implements Queue {
             }
             const singleIngestionTimer = new Date()
             await this.saveEvent(event)
-            resolveOffset(event.kafka_offset!)
+            const offset = event?.uuid ? offsetMap.get(event.uuid) : null
+            if (offset) {
+                resolveOffset(offset)
+                offsetMap.delete(event.uuid!)
+            }
             await heartbeat()
             await commitOffsetsIfNecessary()
             this.pluginsServer.statsd?.timing('kafka_queue.single_ingestion', singleIngestionTimer)
         }
+        for (const removedOffset of offsetMap.values()) {
+            resolveOffset(removedOffset)
+            await heartbeat()
+            await commitOffsetsIfNecessary()
+        }
```

Comment:
This changes almost nothing from the current situation, as it only resolves offsets of events that were discarded after the ones that were not discarded. It doesn't make much sense to resolve offsets completely out of order like this, only the last `resolveOffset` in this will even actually matter. We might as well remove all the `resolveOffset` calls anywhere in this method and just do it at the end with the last offset.
The reason for doing `resolveOffset` is to keep track of processed events INSIDE the `for (const event of processedEvents)` loop, so that if we have, say, 5 events, and for some reason the plugin server exits while processing event no. 3 and only manages to save 1 and 2 to CH, it will know to restart from event no. 3. The problem is, if in the current case if event no. 2 is discarded, its offset won't be resolved, and after restarting the server will restart from event no. 2, in effect processing it twice.
This change _doesn't_ improve that. It only fixes the problem in between batches if the last n events of the batch are discarded, which #123 handles too, except more deliberately.

Reply:
So actually we should just enable `eachBatchAutoResolve: true`? From the docs:

> `eachBatchAutoResolve` configures auto-resolve of batch processing. If set to true, KafkaJS will automatically commit the last offset of the batch if `eachBatch` doesn't throw an error. Default: true.

Am I missing something? I see you had a long comment there about some dragons ahead. 

The way understand it, If we crash, it'll commit whatever was marked with `resolveOffset`. Otherwise it'll always commit the last offset of the batch, even if the last event was discarded. Sounds like what we need? In the worst case, when crashing at event "6" in the sequence "1", "2", "skip", "skip", "skip", "6", it'll retry events 3-5... but it's highly likely they will be skipped again... and I'm OK with more straightforward code here vs more accurate skipping. In practice, most batches contain just an event or two anyway.

Both #123 and this implemented one thing wrong though. The events might be returned from `await this.processEventBatch` in any order (and indeed with random events added to the mix). They could just be reversed. I now added some logic to ingest the events in the original order.

Reply:
Oh and holy macaroni, all tests are green??